scala/private/rule_impls.bzl

```diff
@@ -61,73 +61,55 @@ def compile_scala(
         internal_plugin_jars = ctx.files._dependency_analyzer_plugin
-        compiler_classpath_jars = transitive_compile_jars
 
-        direct_jars = _join_path(cjars.to_list())
+        current_target = str(target_label)
+        optional_scalac_args_map["CurrentTarget"] = current_target
 
+    if dependency_info.need_indirect_info:
         transitive_cjars_list = transitive_compile_jars.to_list()
         indirect_jars = _join_path(transitive_cjars_list)
         indirect_targets = ",".join([str(labels[j.path]) for j in transitive_cjars_list])
 
-        current_target = str(target_label)
-
-        optional_scalac_args = """
-DirectJars: {direct_jars}
-IndirectJars: {indirect_jars}
-IndirectTargets: {indirect_targets}
-CurrentTarget: {current_target}
-        """.format(
-            direct_jars = direct_jars,
-            indirect_jars = indirect_jars,
-            indirect_targets = indirect_targets,
-            current_target = current_target,
-        )
-
-    elif unused_dependency_checker_mode != "off":
-        dependency_analyzer_plugin = ctx.attr._dependency_analyzer_plugin
-        plugins = depset(transitive = [plugins, dependency_analyzer_plugin.files])
-        internal_plugin_jars = ctx.files._dependency_analyzer_plugin
-
-        cjars_list = cjars.to_list()
-        direct_jars = _join_path(cjars_list)
-        direct_targets = ",".join([str(labels[j.path]) for j in cjars_list])
+        optional_scalac_args_map["IndirectJars"] = indirect_jars
+        optional_scalac_args_map["IndirectTargets"] = indirect_targets
 
+    if dependency_info.unused_deps_mode != "off":
         ignored_targets = ",".join([str(d) for d in unused_dependency_checker_ignored_targets])
+        optional_scalac_args_map["UnusedDepsIgnoredTargets"] = ignored_targets
 
-        current_target = str(target_label)
-
-        optional_scalac_args = """
-DirectJars: {direct_jars}
-DirectTargets: {direct_targets}
-IgnoredTargets: {ignored_targets}
-CurrentTarget: {current_target}
-        """.format(
-            direct_jars = direct_jars,
-            direct_targets = direct_targets,
-            ignored_targets = ignored_targets,
-            current_target = current_target,
-        )
-    if is_dependency_analyzer_off(ctx) and not is_plus_one_deps_off(ctx):
-        compiler_classpath_jars = transitive_compile_jars
+    if dependency_info.need_direct_info:
+        cjars_list = cjars.to_list()
+        if dependency_info.need_direct_jars:
+            direct_jars = _join_path(cjars_list)
+            optional_scalac_args_map["DirectJars"] = direct_jars
+        if dependency_info.need_direct_targets:
+            direct_targets = ",".join([str(labels[j.path]) for j in cjars_list])
+            optional_scalac_args_map["DirectTargets"] = direct_targets
+
+    optional_scalac_args = "\n".join([
+        "{k}: {v}".format(k = k, v = v)
+        for (k, v) in sorted(optional_scalac_args_map.items())
```

Comment:
clarify in comment that `sorted` is for input stability and reproducibility?

Reply:
Sounds good